core/src/main/java/bisq/core/account/sign/SignedWitnessService.java

```diff
@@ -0,0 +1,237 @@
+        }
+    }
+
+    private boolean verifyWithECKey(SignedWitness signedWitness) {
+        try {
+            String message = Utilities.encodeToHex(signedWitness.getWitnessHash());
+            String signatureBase64 = new String(signedWitness.getSignature(), Charsets.UTF_8);
+            ECKey key = ECKey.fromPublicOnly(signedWitness.getSignerPubKey());
+            if (arbitratorManager.isPublicKeyInList(Utilities.encodeToHex(key.getPubKey()))) {
+                key.verifyMessage(message, signatureBase64);
+                return true;
+            } else {
+                log.warn("Provided EC key is not in list of valid arbitrators.");
+                return false;
+            }
+        } catch (SignatureException e) {
+            log.warn("verify signedWitness failed. signedWitness={}", signedWitness);
+            return false;
+        }
+    }
+
+    private boolean verifyWithDSAKey(SignedWitness signedWitness) {
+        try {
+            PublicKey signaturePubKey = Sig.getPublicKeyFromBytes(signedWitness.getSignerPubKey());
+            Sig.verify(signaturePubKey, signedWitness.getWitnessHash(), signedWitness.getSignature());
+            return true;
+        } catch (CryptoException e) {
+            log.warn("verify signedWitness failed. signedWitness={}", signedWitness);
+            return false;
+        }
+    }
+
+
+    public Set<SignedWitness> getSignedWitnessSet(AccountAgeWitness accountAgeWitness) {
+        return signedWitnessMap.values().stream()
+                .filter(e -> Arrays.equals(e.getWitnessHash(), accountAgeWitness.getHash()))
+                .collect(Collectors.toSet());
+    }
+
+    // SignedWitness objects signed by arbitrators
+    public Set<SignedWitness> getArbitratorsSignedWitnessSet(AccountAgeWitness accountAgeWitness) {
+        return signedWitnessMap.values().stream()
+                .filter(SignedWitness::isSignedByArbitrator)
+                .filter(e -> Arrays.equals(e.getWitnessHash(), accountAgeWitness.getHash()))
+                .collect(Collectors.toSet());
+    }
+
+    // SignedWitness objects signed by any other peer
+    public Set<SignedWitness> getTrustedPeerSignedWitnessSet(AccountAgeWitness accountAgeWitness) {
+        return signedWitnessMap.values().stream()
+                .filter(e -> !e.isSignedByArbitrator())
+                .filter(e -> Arrays.equals(e.getWitnessHash(), accountAgeWitness.getHash()))
+                .collect(Collectors.toSet());
+    }
+
+    // We go one level up by using the signer Key to lookup for SignedWitness objects which contain the signerKey as
+    // witnessOwnerPubKey
+    public Set<SignedWitness> getSignedWitnessSetBySignerPubKey(byte[] signerPubKey) {
+        return signedWitnessMap.values().stream()
+                .filter(e -> Arrays.equals(e.getWitnessOwnerPubKey(), signerPubKey))
+                .collect(Collectors.toSet());
+    }
+
+    public boolean isValidAccountAgeWitness(AccountAgeWitness accountAgeWitness) {
```

Comment:
There need to be a break in case of circular signing. If someone creates two witnesses that sign each other I think this recursion would go into an infinite loop.

Reply:
Hm... not sure if that can happen. Can you describe the case? If Alice signs Bob and Bob signs Carol and Carol signs Bob it is not an issue as the witness objects are the relevant objects which build the chain, not the signers. But in the lookup for the signers witness we need to probably use the date to use the oldest.

It also does not support yet handling of mulitple signers and when you are allowed to sign...

Reply:
Alice signs Bob's witness and Bob signs Alice's witness.

First level, trying to validaty Bob's witness, we find that it was signed by Alice. We then try to validate Alice's witness and during the second recursion we find that it's signed by Bob and will then try to validate it. Back to where we started in an infinite recursion.

Some leeway has to be given to the time stamps so an attack could be to create this pair and set the same date for both signedWitnesses.

I might well be missing something, and I don't think that's a risk during normal operation, just good to be safe and not add any attack vectors.

Reply:
I think we need to pass through the list and remove items once processed.